### PR TITLE
Add new `Dataset Gallery` docs page to summarize metadata for downloadable datasets

### LINKED ISSUE #5784: Add file and instance metadata to documentation for meshes in `pyvista.examples`
### Describe the feature you would like to be added.

There are hundreds of meshes available for use and/or download in `pyvista.examples`. However, when trying to find a specific kind of mesh to use for testing, generating documentation, or experimentation, it's difficult to know which mesh may be suitable for the task. E.g. perhaps a point cloud or surface mesh is needed, either with many points/cells or with few. Maybe several image volumes are needed, or maybe 2D images for textures. It's also helpful to know if the example includes array data. Since some downloadable meshes may be hundreds of MB in size, it would be helpful to know the file size beforehand.

Ideally, this feature would:

- Add file info to the documentation (e.g. file extension, file size)
- Add mesh instance info to the documentation (e.g. mesh type, num cells, num points, num arrays)
- Summarize this info in a single table for easy lookup/reference (instead of inside individual docstrings)
- BONUS: Add a small preview image of the example in the table

### Links to VTK Documentation, Examples, or Class Definitions.

A custom table could be generated using similar code in [make_tables.py](https://github.com/pyvista/pyvista/blob/3068f5c54de10da55fa16824939666c60bf7f844/doc/source/make_tables.py#L5), for example, which is already used to generate the [Color scheme table](https://docs.pyvista.org/version/stable/api/plotting/charts/plot_color_schemes.html#id1)

### Pseudocode or Screenshots

_No response_

### PR DESCRIPTION
**Pull request by user27182**
### Overview

Continuation of #5814. Resolves #5784.

This PR adds a new Dataset Gallery page to summarize dataset instance and file metadata for datasets. Datasets are also grouped into separate galleries based on their metadata to allow for some filtering and enable the discovery of datasets with specific properties.

In addition, the 'See Also' sections of all datasets from `downloads.py` are updated with cross-reference links to each respective entry in the Dataset Gallery.

Check out a recent preview of the new Dataset Gallery [here](https://661039f500c5bf28c4091d12--meek-duckanoo-d2f5ec.netlify.app/api/examples/dataset_gallery).

### Details
The galleries are created with the following steps.

- The metadata is scrapped using the `dataset_loader` class from #5787. Many new props and bug fixes are added to this class in this PR.
- RST of the metadata fields is then generated for each dataset and stuffed into a [sphinx-design card](https://sphinx-design.readthedocs.io/en/latest/cards.html)
- Cards are collected together to create a [`card-carousel`](https://sphinx-design.readthedocs.io/en/latest/cards.html#card-carousels)
- The rst for each carousel has a similar structure to a table where each row is a card, and so a `DatasetGalleryCarousel` class is added to `make_tables.py` which inherits from the `DocTable` class.
- Datasets are grouped together using filters (e.g. filter dataset type) to create many card carousels
- A `<name>_carousel.rst` file is automatically generated for each carousel. These rst files can be included anywhere in the docs. 
- Card carousels are then grouped together using [tabs](https://sphinx-design.readthedocs.io/en/latest/tabs.html) and [dropdowns](https://sphinx-design.readthedocs.io/en/latest/dropdowns.html)
- The dropdowns and tab groups are defined in a new `dataset_gallery.rst`
- Since there are hundreds of datasets and many dataset categories, using dropdowns, tabs, and card carousels helps keep things organized and compact without the need to scroll vertically past all the datasets. 

Some notable features of the gallery:
- The card carousels were built to accommodate both small screens and large screens, so that the gallery is viewable from mobile or desktop. (If on desktop, try this out by shrinking your browser's width).
- The structure, content, and language used for the dataset types featured in the Dataset Gallery (e.g. PointSets, Gridded Data) aims to closely match the [core API docs](https://docs.pyvista.org/version/stable/api/core/) so that no confusing terminology is introduced.
- Datasets can be given a tag/badge based on their inclusion in a particular carousel. This is used to indicate that a dataset is featured in multiple locations in the gallery.
- New galleries can easily be added, e.g. the medical dataset gallery is defined with just 3 properties and a list of datasets:
https://github.com/pyvista/pyvista/blob/83fa1c7ac1d92ee25b8c332e3122560ad977884d/doc/source/make_tables.py#L1630-L1656

### CODE REVIEW

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
pre-commit.ci autofix

**Comment by user27182** [comment]
pre-commit.ci autofix

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
Other than a few formatting tweak I'm happy with the new layout. 

For the cross refs to the gallery from the doc string See Also sections in the downloads module, I was hoping it would be possible for the cross refs to directly link to the respective card in the gallery. E.g. the See Also for download_cow would link directly to the Cow Dataset card in the gallery. 

But,these links don't work if the drop down menu for that card is collapsed, which it is by default. (You can try this yourself currently by clicking the card badges). 

So, I'm not sure what to do to make these links work. Maybe the all-datasets drop down could be placed at the bottom of the page; instead of at the top where it currently is? And then make that drop-down menu open by default so that the links work?

@tkoyama010 not sure if you have any thoughts on this.

**Comment by user27182** [comment]
I think I'll just try keeping the first drop down where it is and set it open by default, see how that goes.

**Comment by tkoyama010** [comment]
> I think I'll just try keeping the first drop down where it is and set it open by default, see how that goes.

Let's release it in this way first. If you are still interested in the bug after release, let's create an issue to track it :)

**Comment by user27182** [comment]
This is how the see also section currently looks for all functions in downloads. The links are broken , which is the bug. 

You're fine with these links as-is for now?

**Comment by user27182** [comment]
I don't think it's good to keep these broken links, so I could revert changes for that file if you want to merge this without the fix. 

Otherwise I'm planning to fix these with this PR, let me know what you prefer

**Comment by tkoyama010** [comment]
> I don't think it's good to keep these broken links, so I could revert changes for that file if you want to merge this without the fix.
> 
> Otherwise I'm planning to fix these with this PR, let me know what you prefer

Sorry. My expression caused a misunderstanding. What I meant to say was let's merge what links are working. The links are not working, let's discuss the fixes in another thread.

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
Might still need a few minor tweaks to fomatting but otherwise the docs LGTM.

I still need to cleanup the main rst `generate` function code, so expect some changes to `make_tables.py` still, but otherwise in terms of features this is ready for initial review.

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

File: pyvista/examples/downloads.py
Comment (user27182) [inline pyvista/examples/downloads.py:5962]: This should say "Cells Nd" with a space but codespell will complain about this, and adding "Nd" to the list of allowed words in `accept.txt` doesn't work. Adding "nd" does work, but then this will cause codespell to miss future typos of "and" or "2nd". This can't be ignored inline either since it's part of a docstring.

So, I think removing the space here as a workaround is a better alternative than adding "nd" to the codespell dictionary.

**Comment by tkoyama010** [comment]
@user27182 Please resolve the conflict, as I cannot determine which branch should be adopted.

**Comment by user27182** [comment]
To make the merge easier I simply kept all changes from this branch for `_dataset_loader` and `_test_dataset_loader`  instead of going through line-by-line. This ended up reverting some of the changes from #5818 

Would you like me to add back those changes and fix all the ruff noqa skips?

**Comment by tkoyama010** [comment]
> To make the merge easier I simply kept all changes from this branch for `_dataset_loader` and `_test_dataset_loader` instead of going through line-by-line. This ended up reverting some of the changes from #5818
> 
> Would you like me to add back those changes and fix all the ruff noqa skips?

Let's leave it as is here and create another PR after merging this.

**Comment by tkoyama010** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
There are a few graphical bugs with `Cell Type` and `File Ext` properties that should probably be fixed:
image

**Comment by user27182** [comment]
Some less critical graphical issues, these are probably OK for now if they're not fixed by this PR:

The data type (red box) should be aligned with the right side. And the reader type could use backticks to format it the same as the other fields.

image

**Comment by user27182** [comment]
@pyvista-bot preview

File: doc/source/images/dataset-gallery/not_available.png
Comment (tkoyama010) [inline doc/source/images/dataset-gallery/not_available.png:1]: If developers define a new download function, will they need to add these images themselves? If so, I would like instructions on how to do so added to the CONTRIBUTION documentation.

**Review by tkoyama010 — CHANGES_REQUESTED** [review]

File: doc/source/images/dataset-gallery/not_available.png
Comment (user27182) [reply doc/source/images/dataset-gallery/not_available.png:1]: I'll update the contributing section.

In the future this could be automated with #5718 , i.e. with that PR any new documentation would require users to add a test image, but the gallery here could then be made to automatically use images from the `doc_image_cache` directly as the thumbnails.

**Comment by user27182** [comment]
@pyvista-bot preview

File: CONTRIBUTING.rst
Comment (tkoyama010) [inline CONTRIBUTING.rst:842]:
```suggestion
1. Add the new carousel class to the the ``CAROUSEL_LIST`` variable defined
   in ``pyvista/doc/source/make_tables.py``. This will enable the rst to be
   auto-generated for the carousel.
```

File: CONTRIBUTING.rst
Comment (tkoyama010) [inline CONTRIBUTING.rst:846]:
```suggestion
1. Update the ``pyvista/doc/source/api/examples/dataset_gallery.rst`` file to
   include the new generated ``<name>_carousel.rst`` file. E.g. to add the
   carousel as a new drop-down item, add the following:
```

**Review by tkoyama010 — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.rst
Comment (tkoyama010) [inline CONTRIBUTING.rst:802]:
```suggestion
If you have multiple related datasets to contribute, or would like to
group any existing datasets together that share similar properties,
the :ref:`dataset_gallery` can easily be extended to feature these
datasets in a new `card carousel<https://sphinx-design.readthedocs.io/en/latest/cards.html#card-carousels>`_.

For example, to add a new ``Instrument`` dataset category to :ref:`dataset_gallery_category`
featuring two datasets (1) :func:`pyvista.examples.download_guitar` and (2)
:func:`pyvista.examples.download_trumpet`, complete the following steps:
```

**Review by tkoyama010 — CHANGES_REQUESTED** [review]

**Comment by user27182** [comment]
@pyvista-bot preview

**Review by tkoyama010 — CHANGES_REQUESTED** [review]

**Comment by tkoyama010** [comment]
@pyvista-bot preview

**Review by tkoyama010 — DISMISSED** [review]
Now I am ready to merge. Finally, once we have reviewed the preview document, we can merge this. The only concern is the large number of image files, but I decided it was worth adding them since they are small per file and make the document much more expressive. Thanks for your creative suggestions!

**Review by tkoyama010 — DISMISSED** [review]

**Comment by tkoyama010** [comment]
pre-commit.ci autofix

**Comment by tkoyama010** [comment]
pre-commit.ci autofix

**Comment by tkoyama010** [comment]
@pyvista-bot preview

**Review by tkoyama010 — APPROVED** [review]

**Comment by user27182** [comment]
Docs look good to me. Still some minor formatting issues, e.g. spacing values should be truncated. I don't think it should block this merge though.
image

And the links to the guitar and trumpet dataset downloads in contributing.rst aren't working. I can address these as part of #5872 if you like.

**Comment by user27182** [comment]
> The only concern is the large number of image files

The thumbnails used and included with this PR are the actual generated images from the build, just scaled down versions. It might be possible to simply change the image paths to the actual doc/build paths instead of using copies stored in doc/source. This would completely eliminate the need to store the images with the pyvista repo.

**Review by tkoyama010 — APPROVED** [review]

**Pull request merged**
